src/org/javarosa/core/model/Safe2014DagImpl.java

```diff
@@ -61,13 +61,20 @@ public void deleteRepeatGroup(FormInstance mainInstance,
          EvaluationContext evalContext, TreeReference deleteRef,
          TreeElement parentElement, TreeElement deletedElement) {
 
-      Set<QuickTriggerable> alreadyEvaluated = triggerTriggerables(
-            mainInstance, evalContext, deleteRef,
-            new HashSet<QuickTriggerable>(0));
-      publishSummary("Deleted", deleteRef, alreadyEvaluated);
-
-      evaluateChildrenTriggerables(mainInstance, evalContext,
-            deletedElement, false, alreadyEvaluated);
+      // After a repeat group has been deleted, the following repeat groups position has changed.
+      // Evaluate triggerables which depend on the position of the following repeats or on their children.
```

Comment:
This makes it sound like it only evaluates certain triggerables. Maybe it should say something more like "evaluate triggerables for all following repeats and their children?" Let me know if I'm just misunderstanding the wording.

Reply:
@lognaturel  At the first glance it looks like it evaluates only directly related triggerables like `position(..)` but `Safe2014DagImpl` is super smart and will also fire all indirectly related ones. For example:
`<bind calculate="concat(/rgwp/houseM/name, /rgwp/houseM/no)" nodeset="/rgwp/houseM/namecalc" type="string"/>`
This one uses `/rgwp/houseM/no` value which explicitly depends on the position. The dag implementation will fire the concatenation calculation after the position calculation for the `/rgwp/houseM/no` node. 

On the other hand, the following 
`<bind calculate="concat(/rgwp/houseM/name, 'some string value')" nodeset="/rgwp/houseM/namecalc" type="string"/>`
won't be fired because the concatenation doesn't rely at all on the position within the parent. 
This is pretty amazing I think.

Anyway, both "evaluate triggerables for all following repeats and their children" and " evaluate triggerables which depend on the position of the following repeats or on their children" may be confusing. Perhaps we should document the above with a new unit test scenario?

Reply:
That's cool! So we ended up with your "ideal" solution after all.

Adding a test and doing your best to clarify the comment both sound good. Perhaps in the comment you could link to the relevant tests.

Reply:
@lognaturel Ayay I must've made a mistake yesterday while debugging this! It's smart but not as smart as I thought. It doesn't re-evaluate same triggerable when a repeat is deleted but it will re-evaluate any other triggerable inside of the repeat. 
So all I said for
`<bind calculate="concat(/rgwp/houseM/name, /rgwp/houseM/no)" nodeset="/rgwp/houseM/namecalc" type="string"/>` 
is correct but it is not for
`<bind calculate="concat(/rgwp/houseM/name, 'some string value')" nodeset="/rgwp/houseM/namecalc" type="string"/>`.

And I know why this happens. DAG goes over each child of each repeat and looks for triggerables that are dependent on the child reference. In this scenario, it finds `calculate="concat(/rgwp/houseM/name, 'some string value')" `. This should be improved because this triggerable depends on `/rgwp/houseM/name` which doesn't depends on a calculation related to the parent reference - `/rgwp/houseM`. 
I can think about fixing it now or raise an issue for this with some information I already have.

Sorry for the confusion. It came out when I created a test for this. It shows how desperate JavaRosa needs tests.
So no ideal solution for now 😞

Reply:
@lognaturel if fixing the above behavior isn't trivial, I can test some way more complex calculations in a repeat and see if there is still no impact on the performance.